src/Ledger/Ledger/Properties.agda

```diff
@@ -82,29 +97,611 @@ instance
+        ... | ((dp' , _) , q≡ , q∈) = to ∈-map ((gaid , gast) , a≡GA , gaid-gast-∈-removed)
+          where
+          a≡GA : a ≡ GovActionDeposit gaid
+          a≡GA = let open ≡-Reasoning in
+            begin
+            a                       ≡⟨ a≡dp ⟩
+            dp                      ≡⟨ cong (proj₁ ∘ proj₂) q≡ ⟩
+            dp'                     ≡⟨ cong proj₁ $ proj₂ (res-singleton''{m = utxoDeps} $ q∈) ⟩
+            GovActionDeposit gaid   ∎
+
+      govSt' : GovState
+      govSt' = filterᵇ Pᵇ govSt
+
+      connect : filterˢ isGADeposit (dom utxoDeps'') ≡ᵉ filterˢ isGADeposit (dom utxoDeps')
+      connect = filter-pres-≡ᵉ $ dom-cong (res-comp-cong (≡ᵉ.sym χ'≡χ))
+
+      f : GovActionID × GovActionState → DepositPurpose
+      f = λ (id , _) → GovActionDeposit id
+
+      main-invariance-lemma :
+          filterˢ isGADeposit (dom utxoDeps) ≡ᵉ fromList (map f govSt)
+          -------------------------------------------------------------------------
+        → filterˢ isGADeposit (dom utxoDeps') ≡ᵉ fromList (map f (filterᵇ Pᵇ govSt))
+
+      main-invariance-lemma (fst , snd) = ⇛ , ⇚
+        where
+        ⇛ : filterˢ isGADeposit (dom utxoDeps')
+            ⊆ fromList (map f (filterᵇ Pᵇ govSt))
+        ⇛ {a} h with from ∈-filter h
+        ... | (aGADep , a∈) with ∈-map⁻ f (from ∈-fromList (fst (to ∈-filter (aGADep , res-comp-domᵐ a∈))))
+        ... | ((tid , gast) , tidb∈govSt , refl) =
+          to ∈-fromList (∈ˡ-map-filter⁺{P? = λ u → ¿ P u ¿}
+                          ((tid , gast) , refl , tidb∈govSt , (res-comp-dom a∈) ∘ ∈-map⁺-∘))
+
+        ⇚ : fromList (map f (filterᵇ Pᵇ govSt))
+              ⊆ filterˢ isGADeposit (dom utxoDeps')
+        ⇚ {a} h with ∈-map⁻ f (from ∈-fromList h)
+        ... | (aid×st , aid×st∈ , refl) with ∈-filter⁻ (λ u → ¿ P u ¿) aid×st∈
+        ... | (aid×st∈govSt , Paid×st) = ∈-filter-resᶜ-dom⁺ isGADeposit {utxoDeps}
+                                         (refl , a∉χ' , (to dom∈ $ proj₂ (from ∈-filter a∈)))
+          where
+          a∈ : a ∈ filterˢ isGADeposit (dom utxoDeps)
+          a∈ = snd $ to ∈-fromList $ to ∈ˡ-map (aid×st , refl , aid×st∈govSt)
+
+          a∉χ' : a ∉ χ'
+          a∉χ' a∈χ' with ∈-map⁻' a∈χ'
+          ... | q , refl , q∈rem = Paid×st (to ∈-map (q , refl , q∈rem))
+
+
+  -- MAIN THEOREM: CHAIN --
+  module _ (tx : Tx) (Γ : LEnv) (cs : ChainState)
+    where
+    open Tx tx renaming (body to txb); open TxBody txb
+    open LEnv Γ renaming (pparams to pp)
+    open ChainState cs
+    open NewEpochState newEpochState
+
+    ls₀ : LState
+    ls₀ = EpochState.ls epochState
+
+    CHAIN-govDepsMatch : {b : Block} {cs' : ChainState}
+      → _ ⊢ cs ⇀⦇ b ,CHAIN⦈ cs'
+      → govDepsMatch ls₀ → govDepsMatch (EpochState.ls (NewEpochState.epochState (ChainState.newEpochState cs')))
+
+    CHAIN-govDepsMatch (CHAIN _ y) = RTC-preserves-inv (λ {c}{s}{sig} → LEDGER-govDepsMatch sig c s) y
```

Comment:
The fact that this proof doesn't need `EPOCH-govDepsMatch` means that `CHAIN` is incorrect. We actually throw the result of `NEWEPOCH` away. Here's a fixed version:
```
  CHAIN :
    let open Block b; open ChainState s; open NewEpochState nes
        open EpochState epochState; open EnactState es
    in
       record { stakeDistrs = calculateStakeDistrs (EpochState.ls (NewEpochState.epochState newEpochState)) }
         ⊢ newEpochState ⇀⦇ epoch slot ,NEWEPOCH⦈ nes
    →  ⟦ slot , constitution .proj₁ .proj₂ , pparams .proj₁ , es ⟧ˡᵉ
         ⊢ ls ⇀⦇ ts ,LEDGERS⦈ ls'
    ────────────────────────────────
    _ ⊢ s ⇀⦇ b ,CHAIN⦈
        record s { newEpochState = record nes { epochState = record epochState { ls = ls'} } }
```

Reply:
Okay, thanks.  Apparently, I didn't fully understand your earlier explanation about why CHAIN must depend on EPOCH.  Thanks for clarifying.  Hopefully this is enough for me to go on... but I'll let you know if I get stuck trying to prove it.